xlators/storage/posix/src/posix-aio.c

```diff
@@ -402,6 +402,151 @@ posix_aio_writev(call_frame_t *frame, xlator_t *this, fd_t *fd,
     return 0;
 }
 
+int
+posix_aio_fsync_complete(struct posix_aio_cb *paiocb, int res, int res2)
+{
+    call_frame_t *frame = NULL;
+    xlator_t *this = NULL;
+    struct iatt prebuf = {
+        0,
+    };
+    struct iatt postbuf = {
+        0,
+    };
+    int _fd = -1;
+    int op_ret = -1;
+    int op_errno = 0;
+    int ret = 0;
+    fd_t *fd = NULL;
+
+    if (!paiocb) {
+        op_ret = -1;
+        op_errno = EINVAL;
+        goto out;
+    }
+
+    frame = paiocb->frame;
+    this = frame->this;
+    prebuf = paiocb->prebuf;
+    fd = paiocb->fd;
+    _fd = paiocb->_fd;
+
+    if (res < 0) {
+        op_ret = -1;
+        op_errno = -res;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_WRITEV_FAILED,
+               "%s(async) failed fd=%d (%d)",
+               (paiocb->iocb.aio_lio_opcode == IO_CMD_FDSYNC ? "fdatasync"
+                                                             : "fsync"),
+               _fd, res);
+        goto out;
+    }
+
+    ret = posix_fdstat(this, fd->inode, _fd, &postbuf);
+    if (ret != 0) {
+        op_ret = -1;
+        op_errno = errno;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSTAT_FAILED,
+               "fstat failed on fd=%d", _fd);
+        goto out;
+    }
+
+    op_ret = res;
+    op_errno = 0;
+
+out:
+    STACK_UNWIND_STRICT(fsync, frame, op_ret, op_errno, &prebuf, &postbuf,
+                        NULL);
+
+    if (paiocb) {
+        if (paiocb->fd)
+            fd_unref(paiocb->fd);
+        GF_FREE(paiocb);
+    }
+
+    return 0;
+}
+
+int
+posix_aio_fsync(call_frame_t *frame, xlator_t *this, fd_t *fd, int32_t datasync,
+                dict_t *xdata)
+{
+    int32_t op_errno = EINVAL;
+    int _fd = -1, ret = -1;
+    struct posix_fd *pfd = NULL;
+    struct posix_aio_cb *paiocb = NULL;
+    struct posix_private *priv = NULL;
+    struct iocb *iocb = NULL;
+
+    VALIDATE_OR_GOTO(frame, err);
+    VALIDATE_OR_GOTO(this, err);
+    VALIDATE_OR_GOTO(fd, err);
+
+    priv = this->private;
+
+    ret = posix_fd_ctx_get(fd, this, &pfd, &op_errno);
+    if (ret < 0) {
+        gf_msg(this->name, GF_LOG_WARNING, op_errno, P_MSG_PFD_NULL,
+               "pfd is NULL from fd=%p", fd);
+        goto err;
+    }
+
+    _fd = pfd->fd;
+
+    paiocb = GF_CALLOC(1, sizeof(*paiocb), gf_posix_mt_paiocb);
+    if (!paiocb) {
+        op_errno = ENOMEM;
+        goto err;
+    }
+
+    paiocb->frame = frame;
+    paiocb->fd = fd_ref(fd);
+    paiocb->_fd = _fd;
+    paiocb->op = GF_FOP_FSYNC;
+
+    paiocb->iocb.data = paiocb;
+    paiocb->iocb.aio_fildes = _fd;
+    paiocb->iocb.aio_lio_opcode = datasync ? IO_CMD_FDSYNC : IO_CMD_FSYNC;
+    paiocb->iocb.aio_reqprio = 0;
+
+    iocb = &paiocb->iocb;
```

Comment:
Many of these lines are common to all FOPs. It would be good to refactor them into a single function. 

This could be  done in a separate patch if you like. You might also want to fix the CLI to prevent toggling the option when volume is in started state in that patch: https://www.spinics.net/lists/gluster-devel/msg27189.html

Reply:
Common init function looks useful indeed.

Reply:
Please also try to send a CLI fix to prevent toggling storage.linux-aio when the volume is in running state.